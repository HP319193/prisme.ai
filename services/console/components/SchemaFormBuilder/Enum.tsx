import {
  DeleteOutlined,
  InfoCircleOutlined,
  PlusOutlined,
} from '@ant-design/icons';
import {
  Button,
  SchemaForm,
  Tooltip,
  UiOptionsSelect,
} from '@prisme.ai/design-system';
import { Input } from 'antd';
import { useTranslation } from 'next-i18next';
import { ChangeEvent, useCallback } from 'react';

interface Value {
  enum?: string[];
  enumNames?: string[];
  'ui:options'?: { select: Omit<UiOptionsSelect['select'], 'options'> };
}
interface EnumProps {
  value: Value;
  onChange: (v: Value) => void;
}
export const Enum = ({ value, onChange }: EnumProps) => {
  const { t } = useTranslation('common');
  const addValue = useCallback(() => {
    const newValue = { ...value, enum: value.enum || [] };
    const index = newValue.enum.push('');
    if (value.enumNames && !value.enumNames[index]) {
      value.enumNames[index] = '';
    }
    onChange(newValue);
  }, [onChange, value]);

  const deleteValue = useCallback(
    (index: number) => () => {
      const newEnum = (value.enum || []).filter((v, k) => k !== index);
      const newEnumNames = (value.enumNames || []).filter(
        (v, k) => k !== index
      );
      const { enum: _enum, enumNames, ...prevValue } = value;
      const newValue: any = prevValue;
      if (newEnum.length) {
        newValue.enum = newEnum;
      }
      if (newEnumNames.length) {
        newValue.enumNames = newEnumNames;
      }

      onChange(newValue);
    },
    [onChange, value]
  );

  const updateValue = useCallback(
    (index: number) =>
      ({ target: { value: v } }: ChangeEvent<HTMLInputElement>) => {
        const newValue = { ...value, enum: value.enum || [] };
        newValue.enum[index] = v;
        onChange(newValue);
      },
    [onChange, value]
  );
  const updateLabel = useCallback(
    (index: number) =>
      ({ target: { value: v } }: ChangeEvent<HTMLInputElement>) => {
        const newEnumNames = [...(value.enumNames || [])];
        newEnumNames[index] = v;
        onChange({
          ...value,
          enumNames: newEnumNames.map((v) => v || ''),
        });
      },
    [onChange, value]
  );

  return (
    <div className="flex-1 flex-col">
      <div className="flex flex-1 flex-row justify-between">
        <div>
          <label className="font-bold">
            {t('schemaForm.builder.property.enum.label')}
          </label>
          <Tooltip
            title={t('schemaForm.builder.property.enum.description')}
            placement="right"
          >
            <button type="button" className="ml-2">
              <InfoCircleOutlined />
            </button>
          </Tooltip>
        </div>
        <Tooltip
          title={t('schemaForm.builder.property.enum.add')}
          placement="left"
        >
          <Button type="button" onClick={addValue} className="-mt-2">
            <PlusOutlined />
          </Button>
        </Tooltip>
      </div>
      <div>
        {(value.enum || []).map((v, k) => (
          <div
            key={k}
            className="relative flex flex-1 flex-row items-center mb-2"
          >
            <Input
              placeholder={t('schemaForm.builder.property.enum.key')}
              value={(value.enumNames || [])[k]}
              onChange={updateLabel(k)}
            />
            <span className="flex m-1">:</span>
            <Input
              placeholder={t('schemaForm.builder.property.enum.value')}
              value={v}
              onChange={updateValue(k)}
            />
            <Tooltip title={t('schemaForm.builder.property.enum.remove')}>
              <button
                type="button"
                className="absolute top-[50%] right-2 text-neutral-500 text-xs"
                onClick={deleteValue(k)}
              >
                <DeleteOutlined />
              </button>
            </Tooltip>
          </div>
        ))}
      </div>
      <SchemaForm
        schema={{
          type: 'boolean',
          title: t('schemaForm.builder.uiOptions.select.hideSearch.title'),
        }}
        buttons={[]}
        initialValues={value}
        onChange={(formVal) => {
          onChange({
            ...value,
            'ui:options': {
              ...(value?.['ui:options'] || {}),
              select: {
                ...(value?.['ui:options']?.select || {}),
                hideSearch: formVal,
              },
            },
          });
        }}
      />
    </div>
  );
};

export default Enum;
