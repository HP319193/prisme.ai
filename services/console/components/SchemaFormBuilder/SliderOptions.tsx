import { DeleteOutlined, PlusOutlined } from '@ant-design/icons';
import { UiOptionsSlider } from '@prisme.ai/design-system';
import { Input, Tooltip } from 'antd';
import { useTranslation } from 'next-i18next';
import { useCallback } from 'react';
import RichTextEditor from '../RichTextEditor';

interface SliderOptionsProps {
  value: Partial<UiOptionsSlider>;
  onChange: (v: UiOptionsSlider) => void;
}

export const SliderOptions = ({ value, onChange }: SliderOptionsProps) => {
  const { t } = useTranslation('common');

  const addStep = useCallback(() => {
    onChange({
      slider: {
        ...value?.slider,
        steps: [
          ...(value?.slider?.steps || []),
          {
            label: '',
            description: '',
            value: 0,
          },
        ],
      },
    });
  }, [onChange, value]);

  const removeStep = useCallback(
    (index: number) => () => {
      onChange({
        slider: {
          ...value?.slider,
          steps: (value?.slider?.steps || []).filter((v, k) => k !== index),
        },
      });
    },
    [onChange, value?.slider]
  );

  const udpateStepValue = useCallback(
    (index: number) => (key: string) => (newValue: string | number) => {
      onChange({
        slider: {
          ...value?.slider,
          steps: (value?.slider?.steps || []).map((item, k) =>
            k === index
              ? {
                  ...item,
                  [key]: newValue,
                }
              : item
          ),
        },
      });
    },
    [onChange, value?.slider]
  );

  return (
    <div>
      <label className="font-bold flex justify-between mb-2">
        {t('schemaForm.builder.uiOptions.slider.steps.label')}
        <div className="flex justify-end">
          <button onClick={addStep}>
            <Tooltip
              title={t('schemaForm.builder.uiOptions.slider.steps.add')}
              placement="left"
            >
              <PlusOutlined />
            </Tooltip>
          </button>
        </div>
      </label>
      <div>
        {(value?.slider?.steps || []).map(
          ({ description, label, value }, key) => (
            <div key={key} className="flex flex-col relative">
              <div className="flex flex-row">
                <div className="flex flex-col w-[30%]">
                  <label htmlFor={`value-${key}`}>
                    {t('schemaForm.builder.uiOptions.slider.steps.value.label')}
                  </label>
                  <Input
                    id={`value-${key}`}
                    type="number"
                    value={value}
                    onChange={({ target: { value } }) =>
                      udpateStepValue(key)('value')(+value)
                    }
                  />
                </div>
                <div className="flex flex-col flex-1 ml-2">
                  <label htmlFor={`label-${key}`}>
                    {t(
                      'schemaForm.builder.uiOptions.slider.steps._label.label'
                    )}
                  </label>
                  <Input
                    id={`label-${key}`}
                    value={label}
                    onChange={({ target: { value } }) =>
                      udpateStepValue(key)('label')(value)
                    }
                  />
                </div>
              </div>
              <div className="mt-2">
                <label>
                  {t(
                    'schemaForm.builder.uiOptions.slider.steps.description.label'
                  )}
                </label>
                <div className="-mt-6">
                  <RichTextEditor
                    value={description}
                    onChange={udpateStepValue(key)('description')}
                  />
                </div>
              </div>
              <button
                onClick={removeStep(key)}
                className="absolute top-0 right-0"
              >
                <Tooltip
                  title={t('schemaForm.builder.uiOptions.slider.steps.remove')}
                  placement="left"
                >
                  <DeleteOutlined />
                </Tooltip>
              </button>
            </div>
          )
        )}
      </div>
    </div>
  );
};

export default SliderOptions;
