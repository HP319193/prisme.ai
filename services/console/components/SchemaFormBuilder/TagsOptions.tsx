import { SchemaForm, UiOptionsTags } from '@prisme.ai/design-system';
import { useTranslation } from 'next-i18next';

interface TagsOptionsProps {
  value: Partial<UiOptionsTags>;
  onChange: (v: UiOptionsTags) => void;
}

export const TagsOptions = ({ value, onChange }: TagsOptionsProps) => {
  const { t } = useTranslation('common');

  return (
    <SchemaForm
      schema={{
        type: 'object',
        properties: {
          tags: {
            type: 'object',
            title: t('schemaForm.builder.uiOptions.tags.title'),
            properties: {
              allowNew: {
                type: 'boolean',
                title: t('schemaForm.builder.uiOptions.tags.allowNew.title'),
                description: t(
                  'schemaForm.builder.uiOptions.tags.allowNew.description'
                ),
              },
              options: {
                type: 'array',
                title: t('schemaForm.builder.uiOptions.tags.options.title'),
                items: {
                  type: 'object',
                  properties: {
                    value: {
                      title: t('schemaForm.builder.uiOptions.tags.value.title'),
                      type: 'string',
                    },
                    label: {
                      title: t('schemaForm.builder.uiOptions.tags.label.title'),
                      type: 'string',
                    },
                    color: {
                      title: t('schemaForm.builder.uiOptions.tags.color.title'),
                      type: 'string',
                      'ui:widget': 'color',
                    },
                  },
                },
              },
            },
          },
        },
      }}
      buttons={[]}
      initialValues={value}
      onChange={onChange}
    />
  );
};

export default TagsOptions;
