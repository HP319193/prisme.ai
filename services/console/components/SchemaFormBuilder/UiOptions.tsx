import { Collapse, Schema } from '@prisme.ai/design-system';
import { useTranslation } from 'next-i18next';
import { useMemo } from 'react';
import SliderOptions from './SliderOptions';
import TagsOptions from './TagsOptions';
import UploadOptions from './UploadOptions';

interface UiOptionsProps {
  value: Schema;
  onChange: (schema: Schema) => void;
}

const WidgetsWithOptions = ['slider', 'tags', 'upload'];

const getOptionsForm = (widget: typeof WidgetsWithOptions[number]) => {
  switch (widget) {
    case 'slider':
      return SliderOptions;
    case 'tags':
      return TagsOptions;
    case 'upload':
      return UploadOptions;
  }
};

export const UiOptions = ({ value, onChange }: UiOptionsProps) => {
  const { t } = useTranslation('common');
  const widget = value?.['ui:widget'];
  const hasOptions =
    widget && typeof widget === 'string' && WidgetsWithOptions.includes(widget);

  const items = useMemo(() => {
    if (!widget || !hasOptions) return null;
    const Component = getOptionsForm(widget);
    if (!Component) return null;
    return [
      {
        label: t('schemaForm.builder.uiOptions.label'),
        content: (
          <Component
            value={value['ui:options'] || {}}
            onChange={(newValue) => {
              onChange({
                ...value,
                ['ui:options']: newValue,
              });
            }}
          />
        ),
      },
    ];
  }, [hasOptions, onChange, t, value, widget]);

  if (!hasOptions || !items) return null;

  return <Collapse items={items} />;
};

export default UiOptions;
