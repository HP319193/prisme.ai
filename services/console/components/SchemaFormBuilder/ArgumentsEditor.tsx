import { useTranslation } from 'next-i18next';
import Properties from './Properties';

export const ArgumentsEditor = ({ value, onChange }: any) => {
  const { t } = useTranslation('workspaces');

  return (
    <div className="ant-input">
      <div className="text-gray pl-4">{t('automations.arguments.title')}</div>
      <Properties value={value} onChange={onChange} />
    </div>
  );
};

export default ArgumentsEditor;
