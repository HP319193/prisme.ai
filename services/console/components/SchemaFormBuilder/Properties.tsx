import { DeleteOutlined } from '@ant-design/icons';
import {
  Button,
  Input,
  Divider,
  Tooltip,
  Collapse,
  Schema,
} from '@prisme.ai/design-system';
import { useTranslation } from 'next-i18next';
import { useCallback, useMemo } from 'react';
import SchemaFormBuilder from './SchemaFormBuilder';

export const Properties = ({ value, onChange }: any) => {
  const { t } = useTranslation('workspaces');
  const update = useCallback(
    (updatedKey: keyof typeof value) => (schema: Schema) => {
      onChange(
        Object.keys(value).reduce(
          (prev, key) => ({
            ...prev,
            [key]: key === updatedKey ? schema : value[key],
          }),
          {}
        )
      );
    },
    [onChange, value]
  );

  const updateKey = useCallback(
    (prevKey: keyof typeof value) => (newKey: string) => {
      onChange(
        Object.keys(value).reduce((prev, key) => {
          return {
            ...prev,
            [prevKey === key ? newKey : key]: value[key],
          };
        }, {})
      );
    },
    [onChange, value]
  );

  const add = useCallback(() => {
    onChange({ ...value, '': { type: 'string' } });
  }, [onChange, value]);

  const remove = useCallback(
    (oldKey: string) => () => {
      onChange(
        Object.keys(value).reduce(
          (prev, key) =>
            key === oldKey ? prev : { ...prev, [key]: value[key] },
          {}
        )
      );
    },
    [onChange, value]
  );

  const items = useMemo(
    () =>
      Object.keys(value || {}).map((key) => ({
        label: (
          <div
            className="relative flex flex-1"
            onClick={(e) => e.stopPropagation()}
          >
            <Input
              label={t('schema.property.name')}
              value={key}
              onChange={({ target: { value } }) => updateKey(key)(value)}
              pattern={/^[a-zA-Z0-9_]+$/.source}
              containerClassName="flex flex-1"
            />
            <Tooltip title={t('schema.property.delete')} placement="left">
              <button
                type="button"
                onClick={remove(key)}
                className="absolute top-[30px] right-3"
              >
                <DeleteOutlined />
              </button>
            </Tooltip>
          </div>
        ),
        content: (
          <div>
            <div className="pl-4 border-l-[1px] border-x-gray-200">
              <SchemaFormBuilder value={value[key]} onChange={update(key)} />
            </div>
            <Divider />
          </div>
        ),
      })),
    [remove, t, update, updateKey, value]
  );

  return (
    <div className="flex flex-1 flex-col mt-[1rem] centered-collapse">
      <Collapse items={items} expandIconPosition="left" />
      <Button onClick={add}>{t('schema.property.add')}</Button>
    </div>
  );
};

export default Properties;
