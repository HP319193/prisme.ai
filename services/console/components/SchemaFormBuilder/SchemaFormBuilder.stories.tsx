import { Story } from '@storybook/react';
import { useState } from 'react';
import SchemaFormBuilder from './SchemaFormBuilder';

const story = {
  title: 'Console/SchemaFormBuilder',
};
export default story;

const Template: Story<any> = () => {
  const [value, setValue] = useState({});
  return <SchemaFormBuilder value={value} onChange={setValue} />;
};

export const Default = Template.bind({});
