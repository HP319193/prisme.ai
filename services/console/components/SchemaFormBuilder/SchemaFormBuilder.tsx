import { Select, Switch, Schema, schemaTypes } from '@prisme.ai/design-system';
import { useTranslation } from 'next-i18next';
import { useCallback, useMemo } from 'react';
import LocalizedInput from '../LocalizedInput';
import Properties from './Properties';

interface SchemaFormBuilderProps {
  value: Schema;
  onChange: (schema: Schema) => void;
}

export const SchemaFormBuilder = ({
  value,
  onChange,
}: SchemaFormBuilderProps) => {
  const { t } = useTranslation('workspaces');
  const update = useCallback(
    (type: string) => (v: any) => {
      const newValue = { ...value };
      if (type === 'type') {
        if (v !== 'array') {
          delete newValue.items;
        }
        if (v !== 'object') {
          delete newValue.properties;
          delete newValue.additionalProperties;
        }
      }
      onChange({
        ...newValue,
        [type]: v,
      });
    },
    [onChange, value]
  );

  const options = useMemo(
    () =>
      schemaTypes.map((value) => ({
        label: t(`schema.types.${value.replace(':', '_')}`),
        value,
      })),
    [t]
  );

  return (
    <div className="flex flex-1 flex-col">
      <LocalizedInput
        value={value.title || ''}
        onChange={update('title')}
        InputProps={{
          label: 'title',
        }}
        iconMarginTop={17}
      />
      <LocalizedInput
        value={value.description || ''}
        onChange={update('description')}
        InputProps={{
          label: 'description',
        }}
        iconMarginTop={17}
      />
      {/*
      // Required is not already available
      <label className="flex text-gray my-4">
        <Switch
          checked={!!value.required}
          onChange={(checked) => update('required')(checked)}
          className="!mr-2"
        />
        {t('schema.property.required')}
      </label>*/}
      <Select
        selectOptions={options}
        label="type"
        value={value.type}
        onChange={update('type')}
      />
      {value.type === 'array' && (
        <div className="flex flex-1 pl-4 border-l-[1px] border-gray-200">
          <SchemaFormBuilder
            value={value.items || {}}
            onChange={update('items')}
          />
        </div>
      )}
      {value.type === 'object' && (
        <div className="flex flex-1 pl-4 border-l-[1px] border-gray-200">
          <Properties
            value={value.properties || {}}
            onChange={update('properties')}
          />
        </div>
      )}
    </div>
  );
};

export default SchemaFormBuilder;
