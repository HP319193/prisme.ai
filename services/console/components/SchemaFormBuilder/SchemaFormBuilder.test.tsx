import SchemaFormBuilder from './SchemaFormBuilder';
import renderer, { act } from 'react-test-renderer';
import { Button, Schema, Select } from '@prisme.ai/design-system';
import Properties from './Properties';

it('should render', () => {
  const schema: Schema = {
    type: 'object',
    properties: {
      string: {
        type: 'string',
      },
      number: {
        type: 'number',
      },
      boolean: {
        type: 'boolean',
      },
      lorem: {
        type: 'object',
        properties: {
          a: {
            type: 'string',
          },
          b: {
            type: 'array',
            items: {
              type: 'array',
              items: {
                type: 'array',
                items: {
                  type: 'string',
                },
              },
            },
          },
        },
      },
    },
  };

  const onChange = jest.fn();
  const root = renderer.create(
    <SchemaFormBuilder value={schema} onChange={onChange} />
  );
  expect(root.toTree()).toMatchSnapshot();
});

it('should add properties in object', () => {
  const schema = {
    type: 'string',
  } as const;
  const onChange = jest.fn();
  const root = renderer.create(
    <SchemaFormBuilder value={schema} onChange={onChange} />
  );

  expect(root.root.findAllByType(Properties).length).toBe(0);

  act(() => {
    root.root.findAllByType(Select)[1].props.onChange('object');
  });

  expect(onChange).toHaveBeenCalledWith({
    type: 'object',
  });

  act(() => {
    root.update(
      <SchemaFormBuilder value={{ type: 'object' }} onChange={onChange} />
    );
  });

  expect(root.root.findAllByType(Properties).length).toBe(1);

  act(() => {
    root.root.findByType(Properties).findByType(Button).props.onClick();
  });
  expect(onChange).toHaveBeenCalledWith({
    type: 'object',
    properties: {
      '': {
        type: 'string',
      },
    },
  });
});

it('should set items in array', () => {
  const schema = {
    type: 'array',
  } as const;
  const onChange = jest.fn();
  const root = renderer.create(
    <SchemaFormBuilder value={schema} onChange={onChange} />
  );

  expect(root.root.findAllByType(SchemaFormBuilder).length).toBe(2);

  act(() => {
    root.root.findAllByType(SchemaFormBuilder)[1].props.onChange({
      type: 'string',
    });
  });

  expect(onChange).toHaveBeenCalledWith({
    type: 'array',
    items: {
      type: 'string',
    },
  });
});

it('should clean items when changing type', () => {
  const onChange = jest.fn();
  const root = renderer.create(
    <SchemaFormBuilder
      value={{ type: 'array', items: { type: 'string' } }}
      onChange={onChange}
    />
  );
  act(() => {
    root.root.findAllByType(Select)[1].props.onChange('string');
  });
  expect(onChange).toHaveBeenCalledWith({
    type: 'string',
  });
});

it('should clean property when changing type', () => {
  const onChange = jest.fn();
  const root = renderer.create(
    <SchemaFormBuilder
      value={{ type: 'object', properties: { a: { type: 'string' } } }}
      onChange={onChange}
    />
  );
  act(() => {
    root.root.findAllByType(Select)[1].props.onChange('string');
  });
  expect(onChange).toHaveBeenCalledWith({
    type: 'string',
  });
});
