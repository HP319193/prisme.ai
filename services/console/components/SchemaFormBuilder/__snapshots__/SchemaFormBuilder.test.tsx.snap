// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
Object {
  "instance": null,
  "nodeType": "component",
  "props": Object {
    "onChange": [MockFunction],
    "value": Object {
      "properties": Object {
        "boolean": Object {
          "type": "boolean",
        },
        "lorem": Object {
          "properties": Object {
            "a": Object {
              "type": "string",
            },
            "b": Object {
              "items": Object {
                "items": Object {
                  "items": Object {
                    "type": "string",
                  },
                  "type": "array",
                },
                "type": "array",
              },
              "type": "array",
            },
          },
          "type": "object",
        },
        "number": Object {
          "type": "number",
        },
        "string": Object {
          "type": "string",
        },
      },
      "required": Array [
        "foo",
      ],
      "type": "object",
    },
  },
  "rendered": Object {
    "instance": null,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <LocalizedInput
          InputProps={
            Object {
              "label": "title",
            }
          }
          iconMarginTop={17}
          onChange={[Function]}
          value=""
        />,
        <LocalizedInput
          InputProps={
            Object {
              "label": "description",
            }
          }
          iconMarginTop={17}
          onChange={[Function]}
          value=""
        />,
        <label
          className="flex text-gray my-4"
        >
          <Switch
            checked={true}
            className="!mr-2"
            onChange={[Function]}
          />
          schema.property.required
        </label>,
        <Select
          label="type"
          onChange={[Function]}
          selectOptions={
            Array [
              Object {
                "label": "schema.types.string",
                "value": "string",
              },
              Object {
                "label": "schema.types.boolean",
                "value": "boolean",
              },
              Object {
                "label": "schema.types.number",
                "value": "number",
              },
              Object {
                "label": "schema.types.object",
                "value": "object",
              },
              Object {
                "label": "schema.types.array",
                "value": "array",
              },
            ]
          }
          value="object"
        />,
        false,
        <div
          className="flex flex-1 pl-4 border-l-[1px] border-gray-200"
        >
          <Properties
            onChange={[Function]}
            value={
              Object {
                "boolean": Object {
                  "type": "boolean",
                },
                "lorem": Object {
                  "properties": Object {
                    "a": Object {
                      "type": "string",
                    },
                    "b": Object {
                      "items": Object {
                        "items": Object {
                          "items": Object {
                            "type": "string",
                          },
                          "type": "array",
                        },
                        "type": "array",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
                "number": Object {
                  "type": "number",
                },
                "string": Object {
                  "type": "string",
                },
              }
            }
          />
        </div>,
      ],
      "className": "flex flex-1 flex-col",
    },
    "rendered": Array [
      Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {
          "InputProps": Object {
            "label": "title",
          },
          "iconMarginTop": 17,
          "onChange": [Function],
          "value": "",
        },
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {
          "InputProps": Object {
            "label": "description",
          },
          "iconMarginTop": 17,
          "onChange": [Function],
          "value": "",
        },
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Switch
              checked={true}
              className="!mr-2"
              onChange={[Function]}
            />,
            "schema.property.required",
          ],
          "className": "flex text-gray my-4",
        },
        "rendered": Array [
          Object {
            "instance": null,
            "nodeType": "component",
            "props": Object {
              "checked": true,
              "className": "!mr-2",
              "onChange": [Function],
            },
            "rendered": null,
            "type": [Function],
          },
          "schema.property.required",
        ],
        "type": "label",
      },
      Object {
        "instance": null,
        "nodeType": "component",
        "props": Object {
          "label": "type",
          "onChange": [Function],
          "selectOptions": Array [
            Object {
              "label": "schema.types.string",
              "value": "string",
            },
            Object {
              "label": "schema.types.boolean",
              "value": "boolean",
            },
            Object {
              "label": "schema.types.number",
              "value": "number",
            },
            Object {
              "label": "schema.types.object",
              "value": "object",
            },
            Object {
              "label": "schema.types.array",
              "value": "array",
            },
          ],
          "value": "object",
        },
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "nodeType": "host",
        "props": Object {
          "children": <Properties
            onChange={[Function]}
            value={
              Object {
                "boolean": Object {
                  "type": "boolean",
                },
                "lorem": Object {
                  "properties": Object {
                    "a": Object {
                      "type": "string",
                    },
                    "b": Object {
                      "items": Object {
                        "items": Object {
                          "items": Object {
                            "type": "string",
                          },
                          "type": "array",
                        },
                        "type": "array",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
                "number": Object {
                  "type": "number",
                },
                "string": Object {
                  "type": "string",
                },
              }
            }
          />,
          "className": "flex flex-1 pl-4 border-l-[1px] border-gray-200",
        },
        "rendered": Array [
          Object {
            "instance": null,
            "nodeType": "component",
            "props": Object {
              "onChange": [Function],
              "value": Object {
                "boolean": Object {
                  "type": "boolean",
                },
                "lorem": Object {
                  "properties": Object {
                    "a": Object {
                      "type": "string",
                    },
                    "b": Object {
                      "items": Object {
                        "items": Object {
                          "items": Object {
                            "type": "string",
                          },
                          "type": "array",
                        },
                        "type": "array",
                      },
                      "type": "array",
                    },
                  },
                  "type": "object",
                },
                "number": Object {
                  "type": "number",
                },
                "string": Object {
                  "type": "string",
                },
              },
            },
            "rendered": Object {
              "instance": null,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Collapse
                    items={
                      Array [
                        Object {
                          "content": <div>
                            <div
                              className="pl-4 border-l-[1px] border-x-gray-200"
                            >
                              <SchemaFormBuilder
                                onChange={[Function]}
                                value={
                                  Object {
                                    "type": "string",
                                  }
                                }
                              />
                            </div>
                            <Divider />
                          </div>,
                          "label": <div
                            className="relative flex flex-1"
                          >
                            <Input
                              containerClassName="flex flex-1"
                              label="schema.property.name"
                              onChange={[Function]}
                              pattern="^[a-zA-Z0-9_]+$"
                              value="string"
                            />
                            <Tooltip
                              placement="left"
                              title="schema.property.delete"
                            >
                              <button
                                className="absolute top-[30px] right-3"
                                onClick={[Function]}
                                type="button"
                              >
                                <DeleteOutlined />
                              </button>
                            </Tooltip>
                          </div>,
                        },
                        Object {
                          "content": <div>
                            <div
                              className="pl-4 border-l-[1px] border-x-gray-200"
                            >
                              <SchemaFormBuilder
                                onChange={[Function]}
                                value={
                                  Object {
                                    "type": "number",
                                  }
                                }
                              />
                            </div>
                            <Divider />
                          </div>,
                          "label": <div
                            className="relative flex flex-1"
                          >
                            <Input
                              containerClassName="flex flex-1"
                              label="schema.property.name"
                              onChange={[Function]}
                              pattern="^[a-zA-Z0-9_]+$"
                              value="number"
                            />
                            <Tooltip
                              placement="left"
                              title="schema.property.delete"
                            >
                              <button
                                className="absolute top-[30px] right-3"
                                onClick={[Function]}
                                type="button"
                              >
                                <DeleteOutlined />
                              </button>
                            </Tooltip>
                          </div>,
                        },
                        Object {
                          "content": <div>
                            <div
                              className="pl-4 border-l-[1px] border-x-gray-200"
                            >
                              <SchemaFormBuilder
                                onChange={[Function]}
                                value={
                                  Object {
                                    "type": "boolean",
                                  }
                                }
                              />
                            </div>
                            <Divider />
                          </div>,
                          "label": <div
                            className="relative flex flex-1"
                          >
                            <Input
                              containerClassName="flex flex-1"
                              label="schema.property.name"
                              onChange={[Function]}
                              pattern="^[a-zA-Z0-9_]+$"
                              value="boolean"
                            />
                            <Tooltip
                              placement="left"
                              title="schema.property.delete"
                            >
                              <button
                                className="absolute top-[30px] right-3"
                                onClick={[Function]}
                                type="button"
                              >
                                <DeleteOutlined />
                              </button>
                            </Tooltip>
                          </div>,
                        },
                        Object {
                          "content": <div>
                            <div
                              className="pl-4 border-l-[1px] border-x-gray-200"
                            >
                              <SchemaFormBuilder
                                onChange={[Function]}
                                value={
                                  Object {
                                    "properties": Object {
                                      "a": Object {
                                        "type": "string",
                                      },
                                      "b": Object {
                                        "items": Object {
                                          "items": Object {
                                            "items": Object {
                                              "type": "string",
                                            },
                                            "type": "array",
                                          },
                                          "type": "array",
                                        },
                                        "type": "array",
                                      },
                                    },
                                    "type": "object",
                                  }
                                }
                              />
                            </div>
                            <Divider />
                          </div>,
                          "label": <div
                            className="relative flex flex-1"
                          >
                            <Input
                              containerClassName="flex flex-1"
                              label="schema.property.name"
                              onChange={[Function]}
                              pattern="^[a-zA-Z0-9_]+$"
                              value="lorem"
                            />
                            <Tooltip
                              placement="left"
                              title="schema.property.delete"
                            >
                              <button
                                className="absolute top-[30px] right-3"
                                onClick={[Function]}
                                type="button"
                              >
                                <DeleteOutlined />
                              </button>
                            </Tooltip>
                          </div>,
                        },
                      ]
                    }
                  />,
                  <Button
                    onClick={[Function]}
                  >
                    schema.property.add
                  </Button>,
                ],
                "className": "flex flex-1 flex-col",
              },
              "rendered": Array [
                Object {
                  "instance": null,
                  "nodeType": "component",
                  "props": Object {
                    "items": Array [
                      Object {
                        "content": <div>
                          <div
                            className="pl-4 border-l-[1px] border-x-gray-200"
                          >
                            <SchemaFormBuilder
                              onChange={[Function]}
                              value={
                                Object {
                                  "type": "string",
                                }
                              }
                            />
                          </div>
                          <Divider />
                        </div>,
                        "label": <div
                          className="relative flex flex-1"
                        >
                          <Input
                            containerClassName="flex flex-1"
                            label="schema.property.name"
                            onChange={[Function]}
                            pattern="^[a-zA-Z0-9_]+$"
                            value="string"
                          />
                          <Tooltip
                            placement="left"
                            title="schema.property.delete"
                          >
                            <button
                              className="absolute top-[30px] right-3"
                              onClick={[Function]}
                              type="button"
                            >
                              <DeleteOutlined />
                            </button>
                          </Tooltip>
                        </div>,
                      },
                      Object {
                        "content": <div>
                          <div
                            className="pl-4 border-l-[1px] border-x-gray-200"
                          >
                            <SchemaFormBuilder
                              onChange={[Function]}
                              value={
                                Object {
                                  "type": "number",
                                }
                              }
                            />
                          </div>
                          <Divider />
                        </div>,
                        "label": <div
                          className="relative flex flex-1"
                        >
                          <Input
                            containerClassName="flex flex-1"
                            label="schema.property.name"
                            onChange={[Function]}
                            pattern="^[a-zA-Z0-9_]+$"
                            value="number"
                          />
                          <Tooltip
                            placement="left"
                            title="schema.property.delete"
                          >
                            <button
                              className="absolute top-[30px] right-3"
                              onClick={[Function]}
                              type="button"
                            >
                              <DeleteOutlined />
                            </button>
                          </Tooltip>
                        </div>,
                      },
                      Object {
                        "content": <div>
                          <div
                            className="pl-4 border-l-[1px] border-x-gray-200"
                          >
                            <SchemaFormBuilder
                              onChange={[Function]}
                              value={
                                Object {
                                  "type": "boolean",
                                }
                              }
                            />
                          </div>
                          <Divider />
                        </div>,
                        "label": <div
                          className="relative flex flex-1"
                        >
                          <Input
                            containerClassName="flex flex-1"
                            label="schema.property.name"
                            onChange={[Function]}
                            pattern="^[a-zA-Z0-9_]+$"
                            value="boolean"
                          />
                          <Tooltip
                            placement="left"
                            title="schema.property.delete"
                          >
                            <button
                              className="absolute top-[30px] right-3"
                              onClick={[Function]}
                              type="button"
                            >
                              <DeleteOutlined />
                            </button>
                          </Tooltip>
                        </div>,
                      },
                      Object {
                        "content": <div>
                          <div
                            className="pl-4 border-l-[1px] border-x-gray-200"
                          >
                            <SchemaFormBuilder
                              onChange={[Function]}
                              value={
                                Object {
                                  "properties": Object {
                                    "a": Object {
                                      "type": "string",
                                    },
                                    "b": Object {
                                      "items": Object {
                                        "items": Object {
                                          "items": Object {
                                            "type": "string",
                                          },
                                          "type": "array",
                                        },
                                        "type": "array",
                                      },
                                      "type": "array",
                                    },
                                  },
                                  "type": "object",
                                }
                              }
                            />
                          </div>
                          <Divider />
                        </div>,
                        "label": <div
                          className="relative flex flex-1"
                        >
                          <Input
                            containerClassName="flex flex-1"
                            label="schema.property.name"
                            onChange={[Function]}
                            pattern="^[a-zA-Z0-9_]+$"
                            value="lorem"
                          />
                          <Tooltip
                            placement="left"
                            title="schema.property.delete"
                          >
                            <button
                              className="absolute top-[30px] right-3"
                              onClick={[Function]}
                              type="button"
                            >
                              <DeleteOutlined />
                            </button>
                          </Tooltip>
                        </div>,
                      },
                    ],
                  },
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "nodeType": "component",
                  "props": Object {
                    "children": "schema.property.add",
                    "onClick": [Function],
                  },
                  "rendered": "schema.property.add",
                  "type": [Function],
                },
              ],
              "type": "div",
            },
            "type": [Function],
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  "type": [Function],
}
`;
